plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieCatalogModule.java

```diff
@@ -58,9 +63,46 @@ public static NessieIcebergClient createNessieIcebergClient(IcebergNessieCatalog
         icebergNessieCatalogConfig.getBearerToken()
                 .ifPresent(token -> builder.withAuthentication(BearerAuthenticationProvider.create(token)));
 
-        return new NessieIcebergClient(builder.build(NessieApiV1.class),
+        // default version is inferred by uri.
+        IcebergNessieCatalogConfig.ClientApiVersion apiVersion = icebergNessieCatalogConfig.getClientAPIVersion()
+                .orElseGet(() -> inferVersionFromURI(icebergNessieCatalogConfig.getServerUri().toString()));
+
+        NessieApiV1 api;
+        switch (apiVersion) {
+            case V1:
+                api = builder.build(NessieApiV1.class);
+                break;
+            case V2:
+                api = builder.build(NessieApiV2.class);
+                break;
+            default:
+                throw new IllegalArgumentException("Unknown API version '" + apiVersion + "'");
+        }
```

Comment:
This can be now: 

```
       NessieApiV1 api = switch (icebergNessieCatalogConfig.getClientAPIVersion()) {
            case V1 -> builder.build(NessieApiV1.class);
            case V2 -> builder.build(NessieApiV2.class);
        };
```